proposals/0000-net-storage-reads-improvements.md

```diff
@@ -0,0 +1,105 @@
+- While it is possible to ask for multiple specific storage keys at the same time, it is not possible to ask for a list of keys that start with a certain prefix. Due to the way FRAME works, storage keys are grouped by "prefix", for example all account balances start with the same prefix. It is thus a common necessity for a light client to obtain the list of all keys (and possibly their values) that start with a specific prefix. This is currently not possible except by performing multiple queries serially that "walk down" the trie.
+
+Once Polkadot and Kusama will have transitioned to `state_version = 1`, which modifies the format of the trie entries, it will be possible to generate Merkle proofs that contain only the hashes of values in the storage. Thanks to this, it is already possible to prove the existence of a key without sending its entire value (only its hash), or to prove that a value has changed or not between two blocks (by sending just their hashes).
+Thus, the only reason why aforementioned issues exist is because the existing networking messages don't give the possibility for the querier to query this. This is what this proposal aims at fixing.
+
+## Detailed Solution design
+The protobuf schema of the networking protocol can be found here: https://github.com/paritytech/substrate/blob/5b6519a7ff4a2d3cc424d78bc4830688f3b184c0/client/network/light/src/schema/light.v1.proto
+
+The proposal is to modify this protocol in this way:
+
+```diff
+@@ -11,6 +11,7 @@ message Request {
+                RemoteReadRequest remote_read_request = 2;
+                RemoteReadChildRequest remote_read_child_request = 4;
+                // Note: ids 3 and 5 were used in the past. It would be preferable to not re-use them.
++               RemoteReadRequestV2 remote_read_request_v2 = 6;
+        }
+ }
+ 
+@@ -48,6 +49,21 @@ message RemoteReadRequest {
+        repeated bytes keys = 3;
+ }
+ 
++message RemoteReadRequestV2 {
++       required bytes block = 1;
++       optional ChildTrieInfo child_trie_info = 2;  // Read from the main trie if missing.
++       repeated Key keys = 3;
++}
++
++message ChildTrieInfo {
++       enum ChildTrieNamespace {
++               DEFAULT = 1;
++       }
++
++       required bytes hash = 1;
++       required ChildTrieNamespace namespace = 2;
++}
++
+ // Remote read response.
+ message RemoteReadResponse {
+        // Read proof. If missing, indicates that the remote couldn't answer, for example because
+@@ -65,3 +81,8 @@ message RemoteReadChildRequest {
+        // Storage keys.
+        repeated bytes keys = 6;
+ }
++
++message Key {
++       required bytes key = 1;
++       optional bool skipValue = 2; // Defaults to `false` if missing
++       optional bool includeDescendants = 3; // Defaults to `false` if missing
```

Comment:
Regarding request being too big, I would find it neat if it was allowed to reply only to part of the query:
- the n first keys when no descendant
- the n first keys and possibly an incomplete number of descendants for the key n
as long as it match the order of the queried content.
(the problemetic of stopping a process in the middle if proof get too big is something that is needed in pvf context too).

This would be up to the full node to decide where to stop.

The end of content is simply the first key (main key or key in the descendant) that is not part of the proof.

Then the light client could send the query again minus these n first keys.
But to handle the case where the prefix is too big to include all descendant in a single response, I would propose to change this so :
includeDescendant: Option<Option<&[u8]>>
None would means don't include descendant, Some(None) would mean inculde descendant starting by the key itself, Some(Some(&[1, 2, 3u8])) would mean include descendant of prefix, starting with `prefix ++ b"123"` with no need to include value for this starting key (but need to prove the next key is indeed following it).
That way we can query any range.